### PR TITLE
Deprecate inplace in Categorical.remove_unused_categories

### LINKED ISSUE #37643: DEPR: inplace dtype-changing methods on Categorical
- [x] `Categorical.remove_unused_categories` https://github.com/pandas-dev/pandas/pull/37918
- [x] `Cateogrical.remove_categories`  #37981 
- [x] `Categorical.add_categories` https://github.com/pandas-dev/pandas/pull/41118
- [x] `Categorical.reorder_categories` https://github.com/pandas-dev/pandas/pull/41133
- [x] `Categorical.rename_categories` #41186
- [x] `Categorical.set_categories` https://github.com/pandas-dev/pandas/pull/41307

The ability to alter an existing array's dtype is a footgun.  We can return a new object without having to make a copy.

### PR DESCRIPTION
**Pull request by OlehKSS**
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Related to #37643

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -1083,8 +1093,7 @@ def remove_unused_categories(self, inplace=False):
         cat._dtype = new_dtype
         cat._codes = coerce_indexer_dtype(inv, new_dtype.categories)
 
-        if not inplace:
-            return cat
```
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py:1087]: this doesnt get changed until we actually enforce the deprecation a few versions from now

File: pandas/tests/arrays/categorical/test_analytics.py [hunk 2]
```diff
@@ -354,7 +354,7 @@ def test_validate_inplace_raises(self, value):
         with pytest.raises(ValueError, match=msg):
             cat.remove_categories(removals=["D", "E", "F"], inplace=value)
 
-        with pytest.raises(ValueError, match=msg):
+        with tm.assert_produces_warning(FutureWarning):
             cat.remove_unused_categories(inplace=value)
```
Comment (jbrockmendel) [inline pandas/tests/arrays/categorical/test_analytics.py:358]: comment inside this block about why we're getting a FutureWarning.

im guessing this actually needs both the raises and the produces_warning, ie.

```
with pytest.raises(ValueError, match=msg):
    with tm.assert_produces_warning(FutureWarning):
        # GH#37918 inplace kwarg deprecated
        cat.remove_unused_categories(inplace=value)
```

**Comment by jbrockmendel** [comment]
This is a very good start.  For the checkbox at the top, since this doesn't handle all the relevant methods, it doesn't close the issue.  Instead write below the checkboxes "xref #37643"

File: pandas/tests/arrays/categorical/test_analytics.py (hunk 2 above)
Comment (OlehKSS) [reply pandas/tests/arrays/categorical/test_analytics.py:358]: Ok

File: pandas/core/arrays/categorical.py (hunk 1 above)
Comment (OlehKSS) [reply pandas/core/arrays/categorical.py:1087]: Ok

**Comment by OlehKSS** [comment]
@jbrockmendel Thanks for a quick review, I have updated this pull request according to your suggestions

**Comment by jreback** [comment]
thanks @OlehKSS happy to have you do the rest ....

**Pull request merged**